Deleting the last remaining line of a buffer crashes Oni

Oni crashes when I delete the last remaining line of a buffer.
"last line" can be a bit ambiguous but I mean when a file has exactly one line, not the last line of an arbitrarily long file.

Version: Onivim 2 (0.5.1-nightly) (Downloaded on December 4th 2020)
Steps to reproduce:
- Create a new buffer
- Write something
- Go back in normal mode and hit dd
- Move or click with the mouse

Here's the [log file](https://github.com/onivim/oni2/files/5646032/oni2_crash.log) of me doing these exact steps when started oni like so:
`oni2 --disable-extensions --disable-syntax-highlighting --disable-configuration --log-file oni2.log`

Thanks for the issue and repro - was able to reproduce this. Got a callstack:
```
[ERROR] [ 7.935s] Oni2.Exception : Exception (Invalid_argument "index out of bounds"):
Raised by primitive operation at file "src/Core/Buffer.re", line 188, characters 40-65
Called from file "src/Feature/Editor/Editor.re", line 1520, characters 9-134
Called from file "src/Feature/Editor/Feature_Editor.re", line 118, characters 25-63
Called from file "src/Store/Features.re", line 79, characters 36-55
Called from file "src/Store/Features.re", line 1365, characters 6-56
Called from file "src/Updater.re", line 10, characters 33-56
Called from file "list.ml", line 121, characters 24-34
Called from file "src/Updater.re", line 7, characters 4-253
Called from file "src/Store.re", line 111, characters 29-61
Called from file "src/UI/UiEvents.re", line 107, characters 8-24
Called from file "src/UI/Mouse.re", line 428, characters 8-33
Called from file "list.ml", line 110, characters 12-15
Called from file "src/Core/App.re", line 268, characters 19-34
Called from file "packages/reason-sdl2/src/sdl2.re", line 823, characters 10-20
Called from file "packages/reason-sdl2/src/sdl2.re", line 269, characters 17-59
Called from file "packages/reason-sdl2/src/sdl2.re", line 269, characters 17-59
Called from file "src/bin_editor/Oni2_editor.re", line 481, characters 11-26
```

Taking a look!

--- a/CHANGES_CURRENT.md
+++ b/CHANGES_CURRENT.md
@@ -22,6 +22,7 @@
 - #2786 - Extensions: Fix download from open-vsx
 - #2785 - Windows: Fix intermittent crash in `Luv.Process.spawn`
 - #2792 - Vim: Handle count for insert mode commands (`i`/`a`, etc) (fixes #809, #2190)
+- #2802 - Editor: Fix crash when moving mouse in empty buffer (fixes #2800)
 
 ### Performance
 

--- a/src/Feature/Editor/Editor.re
+++ b/src/Feature/Editor/Editor.re
@@ -1502,14 +1502,16 @@ let getCharacterUnderMouse = editor => {
            editor,
          );
 
-       let bufferLine =
-         EditorBuffer.line(
-           EditorCoreTypes.LineNumber.toZeroBased(bytePosition.line),
-           editor.buffer,
-         );
-       if (BufferLine.lengthInBytes(bufferLine)
-           > ByteIndex.toInt(bytePosition.byte)) {
-         byteToCharacter(bytePosition, editor);
+       let lineIdx =
+         EditorCoreTypes.LineNumber.toZeroBased(bytePosition.line);
+       if (lineIdx < EditorBuffer.numberOfLines(editor.buffer)) {
+         let bufferLine = EditorBuffer.line(lineIdx, editor.buffer);
+         if (BufferLine.lengthInBytes(bufferLine)
+             > ByteIndex.toInt(bytePosition.byte)) {
+           byteToCharacter(bytePosition, editor);
+         } else {
+           None;
+         };
        } else {
          None;
        };

--- a/test/Feature/Editor/EditorTests.re
+++ b/test/Feature/Editor/EditorTests.re
@@ -259,4 +259,21 @@ describe("Editor", ({describe, _}) => {
       );
     })
   });
+
+  describe("getCharacterUnderMouse", ({test, _}) => {
+    test("#2800 - should not crash on empty buffer", ({expect, _}) => {
+      let (editor, _buffer) = create([||]);
+
+      // Mouse needs to be moved to reproduce #2800
+      let editor' =
+        editor
+        |> Editor.mouseMove(
+             ~time=Revery.Time.zero,
+             ~pixelX=500.,
+             ~pixelY=500.,
+           );
+
+      expect.equal(Editor.getCharacterUnderMouse(editor'), None);
+    })
+  });
 });

